sound/soc/sof/topology.c

```diff
@@ -1385,67 +1375,167 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
+	if (ret != 0) {
+		dev_err(sdev->dev, "error: parse fir.cfg tokens failed %d\n",
 			le32_to_cpu(private->size));
 		goto err;
 	}
 
-	dev_dbg(sdev->dev, "eq iir %s created\n", swidget->widget->name);
+	sof_dbg_comp_config(scomp, &fir->config);
+
+	swidget->private = (void *)fir;
+
+	ret = sof_ipc_tx_message(sdev->ipc, fir->comp.hdr.cmd, fir,
+				 ipc_size, r, sizeof(*r));
+	if (ret >= 0)
+		return ret;
+err:
+	kfree(fir);
+	return ret;
+}
+
+static int sof_effect_iir_load(struct snd_soc_component *scomp, int index,
+			       struct snd_sof_widget *swidget,
+			       struct snd_soc_tplg_dapm_widget *tw,
+			       struct sof_ipc_comp_reply *r)
+
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_tplg_private *private = &tw->priv;
+	struct sof_ipc_comp_eq_iir *iir;
+	size_t ipc_size = 0, iir_data_size = 0;
+	void *iir_data_p;
+	int ret;
+
+	iir_data_size = le32_to_cpu(private->size) - SOF_EFFECT_DATA_SIZE;
+	ipc_size = sizeof(struct sof_ipc_comp_eq_iir) + iir_data_size;
+	iir_data_p = private->data + SOF_EFFECT_DATA_SIZE;
+
+	iir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!iir)
+		return -ENOMEM;
+
+	/* configure mixer IPC message */
+	iir->comp.hdr.size = ipc_size;
+	iir->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
+	iir->comp.id = swidget->comp_id;
+	iir->comp.type = SOF_COMP_EQ_IIR;
+	iir->comp.pipeline_id = index;
 
-	sof_dbg_comp_config(scomp, &eq->config);
+	memcpy(&iir->data, iir_data_p, iir_data_size);
 
-	swidget->private = (void *)eq;
+	iir->size = iir_data_size;
 
-	ret = sof_ipc_tx_message(sdev->ipc, eq->comp.hdr.cmd, eq,
-				 sizeof(*eq), r, sizeof(*r));
+	ret = sof_parse_tokens(scomp, &iir->config, comp_tokens,
+			       ARRAY_SIZE(comp_tokens), private->array,
+			       SOF_EFFECT_DATA_SIZE);
+	if (ret != 0) {
+		dev_err(sdev->dev, "error: parse iir.cfg tokens failed %d\n",
+			le32_to_cpu(private->size));
+		goto err;
+	}
 
+	sof_dbg_comp_config(scomp, &iir->config);
+
+	swidget->private = (void *)iir;
+
+	ret = sof_ipc_tx_message(sdev->ipc, iir->comp.hdr.cmd, iir,
+				 ipc_size, r, sizeof(*r));
 	if (ret >= 0)
 		return ret;
 err:
-	kfree(eq);
+	kfree(iir);
 	return ret;
 }
 
+/*
+ * Effect Topology
+ */
+
+static int sof_widget_load_effect(struct snd_soc_component *scomp, int index,
+				  struct snd_sof_widget *swidget,
+				  struct snd_soc_tplg_dapm_widget *tw,
+				  struct sof_ipc_comp_reply *r)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_tplg_private *private = &tw->priv;
+	struct sof_ipc_comp_effect config;
+	int ret;
+
+	/* check we have some tokens - we need at least effect type */
+	if (le32_to_cpu(private->size) == 0) {
+		dev_err(sdev->dev, "error: effect tokens not found\n");
+		return -EINVAL;
+	}
+
+	memset(&config, 0, sizeof(config));
+
+	/* get any common EFFECT tokens */
+	ret = sof_parse_tokens(scomp, &config, effect_tokens,
+			       ARRAY_SIZE(effect_tokens), private->array,
+			       SOF_EFFECT_DATA_SIZE);
```

Comment:
I think the last argument could just be le32_to_cpu(private->size) just like in every other case. @juimonen what do you think?

Reply:
if I understood correctly, sof_parse_tokens doesn't like if the size spans the whole priv struct. It assumes it's all vendor token arrays and tries to read the following bytes field as tokens (will fail). So this is limiting the amount and making this actually work.